instructions/src/lib.rs

```diff
@@ -1,731 +1,1154 @@
-pub trait Instruction: Sized {
-    fn into_bytes(self) -> Vec<u8> {
-        let mut buffer = Vec::with_capacity(8);
-        buffer.push(self.opt_code_byte());
-        buffer.push(self.src() << 4 | self.dst());
-        buffer.push((self.offset() & 0x00_ff) as u8);
-        buffer.push(((self.offset() & 0xff_00) >> 8) as u8);
-        buffer.push((self.imm() & 0x00_00_00_ff) as u8);
-        buffer.push(((self.imm() & 0x00_00_ff_00) >> 8) as u8);
-        buffer.push(((self.imm() & 0x00_ff_00_00) >> 16) as u8);
-        buffer.push(((self.imm() & 0xff_00_00_00) >> 24) as u8);
-        buffer
-    }
+use std::fmt;
+
+//the copy of rbpf::ebpf::Insn
+#[derive(Debug, PartialEq)]
+pub struct Insn {
+    pub opc: u8,
+    pub dst: u8,
+    pub src: u8,
+    pub off: i16,
+    pub imm: i32,
+}
 
+pub trait Instruction: Sized {
     fn opt_code_byte(&self) -> u8;
 
     fn dst(&self) -> u8;
 
     fn src(&self) -> u8;
 
-    fn offset(&self) -> u16;
+    fn offset(&self) -> i16;
 
-    fn imm(&self) -> u32;
+    fn imm(&self) -> i32;
 
     fn dst_reg(self, dst: u8) -> Self;
 
     fn src_reg(self, src: u8) -> Self;
 
-    fn offset_bytes(self, offset: u16) -> Self;
+    fn offset_bytes(self, offset: i16) -> Self;
+
+    fn immediate(self, imm: i32) -> Self;
+}
+
+pub trait Assemble {
+    type Asm;
+
+    fn assemble(self) -> Self::Asm;
+}
 
-    fn immediate(self, imm: u32) -> Self;
+pub trait Disassemble {
+    fn disassemble(&self) -> String;
+}
+
+impl<'i, I: Instruction> Assemble for &'i I {
+    type Asm = Vec<u8>;
+
+    fn assemble(self) -> Self::Asm {
+        let mut buffer = Vec::with_capacity(8);
+        buffer.push(self.opt_code_byte());
+        buffer.push(self.src() << 4 | self.dst());
+        buffer.push((self.offset() & 0x00_ff) as u8);
+        buffer.push(((self.offset() & 0xff_00) >> 8) as u8);
+        buffer.push((self.imm() & 0x00_00_00_ff) as u8);
+        buffer.push(((self.imm() & 0x00_00_ff_00) >> 8) as u8);
+        buffer.push(((self.imm() & 0x00_ff_00_00) >> 16) as u8);
+        buffer.push(((self.imm() & 0xff_00_00_00) >> 24) as u8);
+        buffer
+    }
 }
 
 pub struct BpfCode {
     instructions: Vec<u8>
 }
 
+//todo: rename to raw bpf code
 impl BpfCode {
+
+//    pub fn parse(src: String) -> Self {
+//      todo: parse string into instructions
+//    }
+
+//    pub fn from_elf(src: Path) -> Self {
+//        todo: create from elf
+//    }
+
+//    pub fn verify(self) -> VerifiedBpfCode {
+//        todo: impl verification
+//    }
+
     pub fn new() -> Self {
         BpfCode { instructions: vec![] }
     }
 
-    pub fn as_bytes(&self) -> &[u8] {
-        self.instructions.as_slice()
+    pub fn mov_add(&mut self, source: Source, arch: Arch) -> Move {
```

Comment:
The `mov_` prefix, here and in the following functions, is misleading to new eBPF users. I understand you refer to the operation class that is used to form the operation code, but I don't think the user should have to care about that. Thoughts?